src/msg/async/frames_v2.cc

```diff
@@ -0,0 +1,191 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "frames_v2.h"
+
+#include <ostream>
+
+#include <fmt/format.h>
+
+namespace ceph::msgr::v2 {
+
+// Zero-pads bufferlist to padded_len.
+static void pad_zero(bufferlist* bl, uint32_t padded_len) {
```

Comment:
I try to avoid non-const reference parameters because of the appearance of passing by value at the call-site and use pointers for mutated (i.e. output and input/output) parameters instead.
That is, between annotating mutability and nullability, I choose mutability.

I don't think there is a single nullable pointer parameter in this PR.

Reply:
I don't think we can add anything new to what has been already said on e.g. [StackExchange in the discussion on non-optional pointers vs non-`const` references](https://softwareengineering.stackexchange.com/a/299041).
Personally, I appreciate the rule "[F.17: For "in-out" parameters, pass by reference to non-`const`](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f17-for-in-out-parameters-pass-by-reference-to-non-const)" of the C++ Core Guidelines.

> That is, between annotating mutability and nullability, I choose mutability.

I would argue that the need to make such a tough choice is often an indication of other [problem: `void`-returning function with output parameters](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f20-for-out-output-values-prefer-return-values-to-output-parameters). A construct like:

```
output = process(input);
```
looks far more readable than:

```
process(input, &output);
```

Reply:
Well, there is no shortage of C++ style guides, each attempting to paper over the various language warts differently.  As noted in that answer, the problem is that C++ doesn't have an `out` or `mut` keyword to go with references.

While
```
output = process(input);
```
is indeed more readable, it incurs a move.  It might seem insignificant if the move is cheap, but it really adds up when applied to every such function across the codebase.

Reply:
> Well, there is no shortage of C++ style guides, each attempting to paper over the various language warts differently.

I think it's because the language is alive and evolves. Recent versions significantly improved the burden on returning...

>  As noted in that answer, the problem is that C++ doesn't have an `out` or `mut` keyword to go with references.

... to a point where absence of such keyword isn't so important anymore for code making use of these improvements.

Two weeks ago Google has dropped the rule mentioned in [the opening posts on StackExchange](https://softwareengineering.stackexchange.com/q/299021) (commit `7a7a2f510efe7d7fc5ea8fbed549ddb31fac8f3e`).

> is indeed more readable, it incurs a move. It might seem insignificant if the move is cheap, but it really adds up when applied to every such function across the codebase.

I believe readability is far, far, far more important than performance **till profiling shows the opposite**. People tend to worry about burning a bunch of cycles usually without an evidence from `perf` or anything similar. The issue of judging whether particular optimization is worth its costs is too complex for the naked eye.

Anyway, I'm not going to block the changes because of a nit. It would be just nice-to-have those `void`-returning functions brought to the `input = process(output)` idiom.